src/io/sb3/fromSb3.ts

```diff
@@ -85,93 +89,104 @@ function getBlockScript(blocks: { [key: string]: sb3.Block }) {
-                    default:
-                      return {
-                        type: "label",
-                        name: part
-                      };
                   }
-                });
-
-                addInput("PROCCODE", { type: "string", value: mutation.proccode });
-                addInput("ARGUMENTS", { type: "customBlockArguments", value: args });
-                addInput("WARP", { type: "boolean", value: mutation.warp === "true" });
-              } else {
-                // In most cases, just copy the shadow block's fields and inputs
-                // into its parent
-                result = {
-                  ...result,
-                  ...translateInputs(blocks[value].inputs),
-                  ...translateFields(blocks[value].fields, blocks[value].opcode)
+                  const asNum = parseFloat(value);
+                  if (!isNaN(asNum)) {
+                    return asNum;
+                  }
+                  return value;
                 };
+
+                switch (part) {
+                  case "%s":
+                  case "%n":
+                    return {
+                      type: "numberOrString",
+                      name: argNames.splice(0, 1)[0],
+                      defaultValue: optionalToNumber(argDefaults.splice(0, 1)[0])
+                    };
+                  case "%b":
+                    return {
+                      type: "boolean",
+                      name: argNames.splice(0, 1)[0],
+                      defaultValue: argDefaults.splice(0, 1)[0] === "true"
+                    };
+                  default:
+                    return {
+                      type: "label",
+                      name: part
+                    };
+                }
+              });
+
+              addInput("PROCCODE", { type: "string", value: mutation.proccode });
+              addInput("ARGUMENTS", { type: "customBlockArguments", value: args });
+              addInput("WARP", { type: "boolean", value: mutation.warp === "true" });
+            } else {
+              // In most cases, just copy the shadow block's fields and inputs
+              // into its parent
+              result = {
+                ...result,
+                ...translateInputs(blocks[value].inputs),
+                ...translateFields(blocks[value].fields, blocks[value].opcode)
+              };
+            }
+          } else {
+            let isBlocks;
+            if (BlockBase.isKnownBlock(block.opcode)) {
+              const { type } = BlockBase.getDefaultInput(block.opcode, inputName) || {};
+              if (type === "blocks") {
+                isBlocks = true;
               }
+            }
+            isBlocks = isBlocks || inputScript.length > 1;
+            if (isBlocks) {
+              addInput(inputName, { type: "blocks", value: inputScript });
             } else {
               addInput(inputName, { type: "block", value: inputScript[0] });
             }
```

Comment:
If I understand correctly, these cases will only be hit when there's an unknown block. Will that ever actually happen?

Reply:
It's possible that we parse unknown blocks (i.e. `!isKnownBlock(block.opcode)`), in the case of a project containing blocks that sb-edit doesn't specifically recognize yet. These cases could be hit if an extension whose blocks are unknown contained a C block. (I believe such can't happen in Scratch yet, because extensions don't support C blocks, but that will probably change [in the future](https://github.com/LLK/scratch-vm/issues/2120).)